src/pymor/algorithms/eigs.py

```diff
@@ -0,0 +1,304 @@
+        assert E.source == E.range
+        assert E.source == A.source
+
+    if b is None:
+        b = A.source.random(seed=seed)
+    else:
+        assert b in A.source
+
+    assert k < n
+    assert l > k
+
+    V, H, f = _arnoldi(A, E, k, b)
+    k0 = k
+    i = 0
+
+    while True:
+        i += 1
+
+        V, H, f = _extend_arnoldi(A, E, V, H, f, l - k)
+
+        ew, ev = spla.eig(H)
+
+        # truncate small imaginary parts
+        ew.imag[np.abs(ew.imag) / np.abs(ew) < 1e-12] = 0
+
+        if which == 'LM':
+            idx = np.argsort(-np.abs(ew))
+        elif which == 'SM':
+            idx = np.argsort(np.abs(ew))
+        elif which == 'LR':
+            idx = np.argsort(-ew.real)
+        elif which == 'SR':
+            idx = np.argsort(ew.real)
+        elif which == 'LI':
+            idx = np.argsort(-np.abs(ew.imag))
+        elif which == 'SI':
+            idx = np.argsort(np.abs(ew.imag))
+
+        k = k0
+        ews = ew[idx]
+        evs = ev[:, idx]
+
+        rres = f.l2_norm()[0] * np.abs(evs[l - 1]) / np.abs(ews)
+
+        # increase k by one in order to keep complex conjugate pairs together
+        if ews[k - 1].imag != 0 and ews[k - 1].imag + ews[k].imag < 1e-12:
+            k += 1
+
+        if np.all(rres[:k] <= tol) or i >= maxiter:
+            break
+
+        # increase k in order to prevent stagnation
+        k = min(l - 1, k + min(np.count_nonzero(rres[:k] <= tol), (l - k) // 2))
+
+        # sort shifts for QR iteration based on their residual
+        shifts = ews[k:l]
+        srres = rres[k:l]
+        idx = np.argsort(-srres)
+        srres = srres[idx]
+        shifts = shifts[idx]
+
+        # don't use converged unwanted Ritz values as shifts
+        shifts = shifts[srres != 0]
+        k += np.count_nonzero(srres == 0)
+        if shifts[0].imag != 0 and shifts[0].imag + ews[1].imag >= 1e-12:
+            shifts = shifts[1:]
+            k += 1
+
+        H, Qs = _qr_iteration(H, shifts)
+
+        V = V.lincomb(Qs.T)
+        f = V[k] * H[k, k - 1] + f * Qs[l - 1, k - 1]
+        V = V[:k]
+        H = H[:k, :k]
+
+    return ews[:k0], V.lincomb(evs[:, :k0].T)
+
+
+def _arnoldi(A, E, l, b):
+    """Compute an Arnoldi factorization.
+
+    Computes matrices :math:`V_l` and :math:`H_l` and a vector :math:`f_l` such that
+
+    .. math::
+        A V_l = V_l H_l + f_l e_l^T.
+
+    Additionally it holds that :math:`V_l^T V_l` is the identity matrix and :math:`H_l`
+    is an upper Hessenberg matrix. If `E` is not `None` it holds
+
+    .. math::
+        E^{-1} A V_l = V_l H_l + f_l e_l^T.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A.
+    E
+        The |Operator| E.
+    l
+        The length of the Arnoldi factorization.
+    b
+        A |VectorArray| which is used as the initial vector for the iteration.
+
+    Returns
+    -------
+    V
+        A |VectorArray| whose columns span an orthogonal basis for :math:`\mathbb{R}^{l}`.
```

Comment:
It's nicer to refer to elements of a `VectorArray` as "vectors" instead of "columns". Also, isn't `V` a basis for R^n with `l` vectors?